### PR TITLE
Release 0.1.0

### PR DESCRIPTION
**Pull request by dalenewby**
# Sync S3 bucket to B2 bucket using `rclone`

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.git
+.gitignore
+LICENSE
+local.env.dist
+README.md
```
Comment (briskt) [inline .dockerignore:5]: Would this be easier?
```suggestion
# Ignore everything
*

# Whitelist required files
!sync-s3-to-b2.sh
```

File: .github/workflows/build-and-publish.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ vars.DOCKER_ORG }}/${{ github.event.repository.name }}
+          tags: |
+            type=ref,event=branch
+            type=ref,event=tag
+            # set latest tag for master branch
```
Comment (briskt) [inline .github/workflows/build-and-publish.yml:28]:
```suggestion
            # set latest tag for main branch
```

File: .github/workflows/build-and-publish.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ vars.DOCKER_ORG }}/${{ github.event.repository.name }}
+          tags: |
```
Comment (briskt) [inline .github/workflows/build-and-publish.yml:25]: I don't remember what this detail is about. It maybe possible to just omit the `tags` object entirely.

File: .gitignore [hunk 4]
```diff
@@ -9,3 +9,5 @@
 node_modules/
 vendor/
 .terraform/
+
+local.env
```
Comment (briskt) [inline .gitignore:13]: This is not necessary with the *.env on line 2

File: README.md [hunk 5]
```diff
@@ -1,2 +1,37 @@
-# template-public
-general template for public repositories
+# sync-s3-to-b2
+Service to synchronize an AWS S3 bucket path to a Backblaze B2 bucket path
+
+This process uses `rclone sync` to copy changes from a specified directory in an AWS S3 bucket to a specified directory in a Backblaze B2 bucket. After configuring `rclone`, the script executes `rclone sync --checksum --create-empty-src-dirs --metadata --transfers 32 ${RCLONE_ARGUMENTS} s3bucket:${S3_BUCKET}/${S3_PATH} b2bucket:${B2_BUCKET}/${B2_PATH}`.
```
Comment (briskt) [inline README.md:4]: Is this level of detail helpful enough to risk becoming out of sync with the code?   Maybe it's enough to say something like "After configuring rclone, the script executes rclone to synchronize the configured buckets."

File: README.md [hunk 6]
```diff
@@ -1,2 +1,37 @@
...
+`S3_PATH` - path within the AWS S3 bucket
+
+`B2_BUCKET` - Backblaze B2 bucket name, e.g., _my-b2-copy-bucket_
+
+`B2_PATH` - path within the Backblaze B2 bucket
+
+`RCLONE_ARGUMENTS` - (optional) Extra rclone arguments. For example, adding `--combined -` to `RCLONE_ARGUMENTS` will "list all file paths with a symbol and then a space and then the path to tell you what happened to it".
+
+## Docker Hub
+This image is built automatically on Docker Hub as [silintl/sync-s3-to-b2](https://hub.docker.com/r/silintl/sync-s3-to-b2/).
```
Comment (briskt) [inline README.md:37]:
```suggestion
This image is built automatically and uploaded to Docker Hub at [silintl/sync-s3-to-b2](https://hub.docker.com/r/silintl/sync-s3-to-b2/).
```
We don't use Docker Hub to build the image. Wouldn't want someone to be confused about this.

File: sync-s3-to-b2.sh [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+MYNAME="sync-s3-to-b2"
+STATUS=0
+
+echo "${MYNAME}: Started"
+
+if [ "${B2_APPLICATION_KEY_ID}" = "" ]; then
+	echo "${MYNAME}: FATAL: environment variable B2_APPLICATION_KEY_ID is required."
+	STATUS=1
+fi
```
Comment (briskt) [inline sync-s3-to-b2.sh:11]: Would a small function help reduce some of the duplication?

**Review by briskt — APPROVED** [review]
no real problems, but some things to consider

File: .dockerignore (hunk 1 above)
Comment (mtompset) [reply .dockerignore:5]: > Would this be easier?

There's a Dockerfile too.

File: .github/workflows/build-and-publish.yml (hunk 2 above)
Comment (mtompset) [reply .github/workflows/build-and-publish.yml:28]: We are using develop and main for newer projects. Using master upsets some people associating it with slavery.

File: .dockerignore (hunk 1 above)
Comment (briskt) [reply .dockerignore:5]: No need to include Dockerfile

File: README.md (hunk 5 above)
Comment (mtompset) [reply README.md:4]: This ties into my validation of environment variables comment indicating a way to improve the script.

File: README.md (hunk 6 above)
Comment (mtompset) [reply README.md:37]: Or just saying it is available on Docker hub, ignoring how it gets there, is another's possibility.

File: sync-s3-to-b2.sh (hunk 7 above)
Comment (mtompset) [reply sync-s3-to-b2.sh:11]: And even an abstraction function to conglomerate the validations to a single line in the main code.

File: .dockerignore (hunk 1 above)
Comment (dalenewby) [reply .dockerignore:5]: It seems odd to me to use a file named `ignore` to list files that you do **not** want to ignore.  If I were to add another file to the Docker image in the future, I'm not convinced that I'd remember to add it to the `.dockerignore`.  To me, ignoring a few files is a buffing/polishing/cleanup activity that saves a few milliseconds during the build.

File: .github/workflows/build-and-publish.yml (hunk 2 above)
Comment (dalenewby) [reply .github/workflows/build-and-publish.yml:28]: And the comment doesn't match the code.  This file was copied from some other project; I didn't look at it closely.

File: .github/workflows/build-and-publish.yml (hunk 3 above)
Comment (dalenewby) [reply .github/workflows/build-and-publish.yml:25]: I don't know what it does either.  It was copied from a different project.

File: .dockerignore (hunk 1 above)
Comment (briskt) [reply .dockerignore:5]: The same could be said for remembering to ignore a file. If you forget to whitelist a file, you know pretty quickly. But this is all splitting hairs for a small project like this.

File: README.md (hunk 5 above)
Comment (dalenewby) [reply README.md:4]: I'd rather read two English sentences that summarize what the code does, even if the included command is out of date, than comprehend 74 lines of shell script only to discover that the `rclone sync` command won't work for me.

Comment (briskt) [reply README.md:4]: I think we're in agreement, then. Keep the English, toss the script command. Right?

Comment (dalenewby) [reply README.md:4]: I don't think we are.  If the `rclone sync` command isn't in the README, then the user is forced to read 74 lines of shell script before he finds out how the synchronization is accomplished.

File: sync-s3-to-b2.sh (hunk 7 above)
Comment (dalenewby) [reply sync-s3-to-b2.sh:11]: I think I'll leave the abstraction function to a future developer.

**Pull request merged**